core/state/trie_prefetcher_extra_test.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/stretchr/testify/require"
+)
+
+const namespace = "chain"
+
+// BenchmarkPrefetcherDatabase first add TEST_DB_KVS key-value pairs to a trie
+// stored in leveldb, then runs benchmarks for updating the trie with different
+// numbers of updates and different prefetcher configuration.
+
+func BenchmarkPrefetcherDatabase(b *testing.B) {
+	require := require.New(b)
+
+	dir := b.TempDir()
+	if env := os.Getenv("TEST_DB_DIR"); env != "" {
+		dir = env
+	}
+	wantKVs := 100_000
+	if env := os.Getenv("TEST_DB_KVS"); env != "" {
+		var err error
+		wantKVs, err = strconv.Atoi(env)
+		require.NoError(err)
+	}
+
+	levelDB, err := rawdb.NewLevelDBDatabase(path.Join(dir, "level.db"), 0, 0, "", false)
+	require.NoError(err)
+
+	root := types.EmptyRootHash
+	count := uint64(0)
+	block := uint64(0)
+
+	rootKey := []byte("root")
+	countKey := []byte("count")
+	blockKey := []byte("block")
+	got, err := levelDB.Get(rootKey)
+	if err == nil {
+		root = common.BytesToHash(got)
+	}
+	got, err = levelDB.Get(countKey)
+	if err == nil {
+		count = binary.BigEndian.Uint64(got)
+	}
+	got, err = levelDB.Get(blockKey)
+	if err == nil {
+		block = binary.BigEndian.Uint64(got)
+	}
+
+	// Make a trie on the levelDB
+	address1 := common.Address{42}
+	address2 := common.Address{43}
+	addBlock := func(db Database, snaps *snapshot.Tree, kvsPerBlock int, prefetchers int) (statedb *StateDB) {
+		statedb, root, err = addKVs(db, snaps, address1, address2, root, block, kvsPerBlock, prefetchers)
+		require.NoError(err)
+		count += uint64(kvsPerBlock)
+		block++
+
+		return statedb
+	}
+
+	lastCommit := block
+	commit := func(levelDB ethdb.Database, snaps *snapshot.Tree, db Database) {
+		require.NoError(db.TrieDB().Commit(root, false))
+
+		for i := lastCommit + 1; i <= block; i++ {
+			require.NoError(snaps.Flatten(fakeHash(i)))
+		}
+		lastCommit = block
+
+		// update the tracking keys
+		err = levelDB.Put(rootKey, root.Bytes())
+		require.NoError(err)
+		err = database.PutUInt64(levelDB, blockKey, block)
+		require.NoError(err)
+		err = database.PutUInt64(levelDB, countKey, count)
+		require.NoError(err)
+	}
+
+	tdbConfig := &triedb.Config{
+		HashDB: &hashdb.Config{
+			CleanCacheSize: 3 * 1024 * 1024 * 1024,
+		},
+	}
+	db := NewDatabaseWithConfig(levelDB, tdbConfig)
+	snaps := snapshot.NewTestTree(levelDB, fakeHash(block), root)
+	for count < uint64(wantKVs) {
+		previous := root
+		_ = addBlock(db, snaps, 100_000, 0) // Note this updates root and count
+		b.Logf("Root: %v, kvs: %d, block: %d", root, count, block)
+
+		// Commit every 10 blocks or on the last iteration
+		if block%10 == 0 || count >= uint64(wantKVs) {
+			commit(levelDB, snaps, db)
+			b.Logf("Root: %v, kvs: %d, block: %d (committed)", root, count, block)
+		}
+		if previous != root {
+			require.NoError(db.TrieDB().Dereference(previous))
+		} else {
+			panic("root did not change")
```

Comment:
Why panic and not `b.Fatal()`? You don't really need the dump that comes with a panic.